File: frontend/lib/src/queryParamSerializers.ts
/**
 * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2025)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Serializers for converting widget values to/from URL query parameter strings.
 *
 * These serializers are used when widgets are bound to query parameters.
 * The format parameter on widgets is for display only and does NOT affect serialization.
 * We always use canonical formats that are unambiguous and parseable.
 */

// Type for a function that serializes a widget value to a query param string
export type QueryParamSerializer<T> = (value: T) => string | string[]

// Type for a function that deserializes a query param string to a widget value
export type QueryParamDeserializer<T> = (
  value: string | string[] | null
) => T | undefined

// --- Boolean Serializers (for Checkbox, Toggle) ---

export function serializeBool(value: boolean): string {
  return value ? "true" : "false"
}

export function deserializeBool(value: string | string[] | null): boolean {
  if (value === null) return false
  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  return ["true", "1", "yes", "on"].includes(str.toLowerCase())
}

// --- String Serializers (for TextInput, TextArea) ---

export function serializeString(value: string | null): string {
  return value ?? ""
}

export function deserializeString(
  value: string | string[] | null
): string | undefined {
  if (value === null) return undefined
  return Array.isArray(value) ? (value[value.length - 1] ?? "") : value
}

// --- Number Serializers (for NumberInput) ---

export function serializeNumber(value: number | null): string {
  if (value === null) return ""

  // For integers, use string directly
  if (Number.isInteger(value)) {
    return String(value)
  }

  // For floats, use toString which gives full precision
  // Then clean up trailing zeros but keep at least one decimal place
  const str = String(value)
  if (str.includes(".") && !str.includes("e")) {
    const parts = str.split(".")
    const decimalPart = parts[1].replace(/0+$/, "") || "0"
    return `${parts[0]}.${decimalPart}`
  }
  return str
}

export function deserializeNumber(
  value: string | string[] | null,
  asInt = false
): number | undefined {
  if (value === null) return undefined

  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  if (!str) return undefined

  const parsed = asInt ? parseInt(str, 10) : parseFloat(str)
  return Number.isNaN(parsed) ? undefined : parsed
}

// --- Color Serializers (for ColorPicker) ---

export function serializeColor(value: string | null): string {
  if (!value) return ""
  // Remove # prefix for cleaner URLs
  return value.replace(/^#/, "").toLowerCase()
}

export function deserializeColor(
  value: string | string[] | null
): string | undefined {
  if (value === null) return undefined

  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  if (!str) return undefined

  // Add # prefix if not present
  return str.startsWith("#") ? str.toLowerCase() : `#${str.toLowerCase()}`
}

// --- Date Serializers (for DateInput) ---

/**
 * Serialize a Date to ISO format (YYYY-MM-DD).
 */
export function serializeDate(value: Date | null): string {
  if (!value || !(value instanceof Date) || Number.isNaN(value.getTime())) {
    return ""
  }
  // Use local date components to avoid timezone issues
  const year = value.getFullYear()
  const month = String(value.getMonth() + 1).padStart(2, "0")
  const day = String(value.getDate()).padStart(2, "0")
  return `${year}-${month}-${day}`
}

/**
 * Deserialize an ISO date string (YYYY-MM-DD) to a Date.
 */
export function deserializeDate(
  value: string | string[] | null
): Date | undefined {
  if (value === null) return undefined

  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  if (!str) return undefined

  // Parse ISO format YYYY-MM-DD
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(str)
  if (!match) return undefined

  const [, yearStr, monthStr, dayStr] = match
  const year = parseInt(yearStr, 10)
  const month = parseInt(monthStr, 10) - 1 // JS months are 0-indexed
  const day = parseInt(dayStr, 10)

  const date = new Date(year, month, day)
  // Validate the date is valid (e.g., not Feb 30)
  if (
    date.getFullYear() !== year ||
    date.getMonth() !== month ||
    date.getDate() !== day
  ) {
    return undefined
  }

  return date
}

/**
 * Serialize a date range to comma-separated ISO format.
 */
export function serializeDateRange(
  value: [Date, Date] | [Date] | Date[] | null
): string {
  if (!value || value.length === 0) return ""

  if (value.length === 1) {
    return serializeDate(value[0])
  }

  // value.length >= 2 at this point
  const start = serializeDate(value[0])
  const end = value[1] ? serializeDate(value[1]) : ""
  if (!start || !end) return ""

  return `${start},${end}`
}

/**
 * Deserialize comma-separated ISO dates to a date range.
 * Returns a tuple of [start, end] dates, or a single [date] for single values.
 */
export function deserializeDateRange(
  value: string | string[] | null
): [Date, Date] | [Date] | undefined {
  if (value === null) return undefined

  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  if (!str) return undefined

  const parts = str.split(",")
  if (parts.length === 1) {
    const date = deserializeDate(parts[0])
    return date ? [date] : undefined
  }

  if (parts.length === 2) {
    const start = deserializeDate(parts[0])
    const end = deserializeDate(parts[1])
    if (start && end) {
      return [start, end]
    }
  }

  return undefined
}

// --- Time Serializers (for TimeInput) ---

/**
 * Serialize a time (as Date or {hour, minute, second}) to ISO format (HH:MM:SS or HH:MM).
 */
export function serializeTime(
  value: Date | { hour: number; minute: number; second?: number } | null
): string {
  if (!value) return ""

  let hour: number
  let minute: number
  let second: number

  if (value instanceof Date) {
    if (Number.isNaN(value.getTime())) return ""
    hour = value.getHours()
    minute = value.getMinutes()
    second = value.getSeconds()
  } else {
    hour = value.hour
    minute = value.minute
    second = value.second ?? 0
  }

  const hh = String(hour).padStart(2, "0")
  const mm = String(minute).padStart(2, "0")

  // Only include seconds if non-zero
  if (second > 0) {
    const ss = String(second).padStart(2, "0")
    return `${hh}:${mm}:${ss}`
  }

  return `${hh}:${mm}`
}

/**
 * Deserialize an ISO time string (HH:MM or HH:MM:SS) to a time object.
 * Returns {hour, minute, second} for use with time widgets.
 */
export function deserializeTime(
  value: string | string[] | null
): { hour: number; minute: number; second: number } | undefined {
  if (value === null) return undefined

  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  if (!str) return undefined

  // Parse HH:MM or HH:MM:SS format
  const match = /^(\d{1,2}):(\d{2})(?::(\d{2}))?$/.exec(str)
  if (!match) return undefined

  const hour = parseInt(match[1], 10)
  const minute = parseInt(match[2], 10)
  const second = match[3] ? parseInt(match[3], 10) : 0

  // Validate ranges
  if (
    hour < 0 ||
    hour > 23 ||
    minute < 0 ||
    minute > 59 ||
    second < 0 ||
    second > 59
  ) {
    return undefined
  }

  return { hour, minute, second }
}

// --- Number Range Serializers (for Slider range mode) ---

/**
 * Serialize a number range to comma-separated format.
 */
export function serializeNumberRange(
  value: [number, number] | number | null
): string {
  if (value === null) return ""

  if (typeof value === "number") {
    return serializeNumber(value)
  }

  if (Array.isArray(value) && value.length === 2) {
    return `${serializeNumber(value[0])},${serializeNumber(value[1])}`
  }

  return ""
}

/**
 * Deserialize comma-separated numbers to a number range or single value.
 */
export function deserializeNumberRange(
  value: string | string[] | null,
  asInt = false
): [number, number] | number | undefined {
  if (value === null) return undefined

  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  if (!str) return undefined

  const parts = str.split(",")

  if (parts.length === 1) {
    // Single value
    return deserializeNumber(parts[0], asInt)
  }

  if (parts.length === 2) {
    // Range value
    const min = deserializeNumber(parts[0], asInt)
    const max = deserializeNumber(parts[1], asInt)
    if (min !== undefined && max !== undefined) {
      return [min, max]
    }
  }

  return undefined
}

// --- Utility: Get last value from array or string ---

export function getLastValue(value: string | string[] | null): string | null {
  if (value === null) return null
  return Array.isArray(value) ? (value[value.length - 1] ?? null) : value
}

// --- Date String Serializers (for DateInput which uses YYYY/MM/DD strings) ---

/**
 * Serialize date strings from DateInput format (YYYY/MM/DD) to ISO format (YYYY-MM-DD).
 * DateInput stores dates as strings internally, so we just convert the separator.
 */
export function serializeDateString(value: string | null): string {
  if (!value) return ""
  // Convert YYYY/MM/DD to YYYY-MM-DD
  return value.replace(/\//g, "-")
}

/**
 * Serialize an array of date strings to comma-separated ISO format.
 * Handles both single dates and date ranges from DateInput.
 */
export function serializeDateStringArray(value: string[] | null): string {
  if (!value || value.length === 0) return ""

  // Convert each date from YYYY/MM/DD to YYYY-MM-DD
  const isoDates = value.map(d => d.replace(/\//g, "-"))
  return isoDates.join(",")
}

/**
 * Deserialize ISO date strings (YYYY-MM-DD) back to DateInput format (YYYY/MM/DD).
 */
export function deserializeDateString(
  value: string | string[] | null
): string | undefined {
  if (value === null) return undefined

  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  if (!str) return undefined

  // Convert YYYY-MM-DD to YYYY/MM/DD
  return str.replace(/-/g, "/")
}

/**
 * Deserialize comma-separated ISO dates to DateInput string array format.
 */
export function deserializeDateStringArray(
  value: string | string[] | null
): string[] | undefined {
  if (value === null) return undefined

  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  if (!str) return undefined

  // Split by comma and convert each YYYY-MM-DD to YYYY/MM/DD
  const parts = str.split(",")
  return parts.map(d => d.replace(/-/g, "/"))
}

// --- Selection Widget Serializers (Radio, Selectbox, Multiselect, SelectSlider) ---

/**
 * Serialize a Radio/Selectbox selection by converting the selected index to its option value.
 * @param index - The selected index (0-based)
 * @param options - The options array from the widget proto
 * @returns The string representation of the selected option, or "" if invalid
 */
export function serializeSelectionIndex(
  index: number | null | undefined,
  options: readonly string[]
): string {
  if (
    index === null ||
    index === undefined ||
    index < 0 ||
    index >= options.length
  ) {
    return ""
  }
  return options[index]
}

/**
 * Deserialize a query param value to a Radio/Selectbox selection index.
 * @param value - The query param value
 * @param options - The options array from the widget proto
 * @returns The index of the matching option, or undefined if not found
 */
export function deserializeSelectionIndex(
  value: string | string[] | null,
  options: readonly string[]
): number | undefined {
  if (value === null) return undefined

  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  if (!str) return undefined

  const index = options.indexOf(str)
  return index >= 0 ? index : undefined
}

/**
 * Serialize a Multiselect selection (array of option values).
 * @param values - Array of selected values (strings)
 * @returns Array of selected option strings
 */
export function serializeMultiselect(values: string[] | null): string[] {
  if (!values || values.length === 0) return []
  return values
}

/**
 * Deserialize query param values to Multiselect selections.
 * @param value - Single string or array of strings from query params
 * @param options - The options array from the widget proto
 * @returns Array of matching option values
 */
export function deserializeMultiselect(
  value: string | string[] | null,
  options: readonly string[]
): string[] | undefined {
  if (value === null) return undefined

  // Handle both single value and array
  const values = Array.isArray(value) ? value : [value]
  if (values.length === 0 || (values.length === 1 && !values[0])) {
    return undefined
  }

  // Only include values that are valid options
  const result = values.filter(v => options.includes(v))
  return result.length > 0 ? result : undefined
}

/**
 * Serialize a SelectSlider selection (single value or range).
 * @param indices - The selected index/indices as numbers
 * @param options - The options array from the widget proto
 * @returns Comma-separated string of option values for ranges, or single value
 */
export function serializeSelectSlider(
  indices: number[] | null,
  options: readonly string[]
): string {
  if (!indices || indices.length === 0) return ""

  const values = indices.map(idx => {
    if (idx < 0 || idx >= options.length) return ""
    return options[idx]
  })

  // Filter out invalid values
  const validValues = values.filter(v => v !== "")
  if (validValues.length === 0) return ""

  if (validValues.length === 1) {
    return validValues[0]
  }

  // Range: comma-separated
  return validValues.join(",")
}

/**
 * Deserialize query param value to SelectSlider indices.
 * @param value - Single string (possibly comma-separated for range) from query params
 * @param options - The options array from the widget proto
 * @returns Array of indices, or undefined if not found
 */
export function deserializeSelectSlider(
  value: string | string[] | null,
  options: readonly string[]
): number[] | undefined {
  if (value === null) return undefined

  const str = Array.isArray(value) ? (value[value.length - 1] ?? "") : value
  if (!str) return undefined

  // Check for comma-separated range
  if (str.includes(",")) {
    const parts = str.split(",")
    const indices = parts.map(p => options.indexOf(p.trim()))
    // All parts must be valid
    if (indices.some(idx => idx < 0)) return undefined
    return indices
  }

  // Single value
  const index = options.indexOf(str)
  return index >= 0 ? [index] : undefined
}

// --- Query Param Key Detection ---

/** Prefix for user keys that should be bound to URL query parameters. */
export const QUERY_PARAM_KEY_PREFIX = "?"

/**
 * Returns true if the user key indicates query param binding.
 * A user key starting with "?" indicates that the widget should be
 * bound to a URL query parameter of the same name (without the prefix).
 */
export function isQueryParamKey(userKey: string | undefined | null): boolean {
  if (!userKey) return false
  return userKey.startsWith(QUERY_PARAM_KEY_PREFIX)
}

/**
 * Extracts the query param name from a user key with the "?" prefix.
 * Assumes the key starts with "?" - use isQueryParamKey() to check first.
 *
 * @param userKey - The user key with "?" prefix (e.g., "?enabled")
 * @returns The query param name (e.g., "enabled")
 */
export function extractQueryParamName(userKey: string): string {
  return userKey.slice(QUERY_PARAM_KEY_PREFIX.length)
}

File: lib/streamlit/runtime/state/query_param_serializers.py
# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2025)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Serializers for converting widget values to/from URL query parameter strings.

These serializers are used when widgets are bound to query parameters. Unlike the
widget's internal serializers (which convert to/from protobuf), these convert
to/from URL-safe string representations.

The format parameter on widgets is for display only and does NOT affect serialization.
We always use canonical formats that are unambiguous and parseable.
"""

from __future__ import annotations

from datetime import date, datetime, time
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Callable, Sequence


def serialize_bool(value: bool) -> str:
    """Serialize a boolean value to a query param string.

    Parameters
    ----------
    value : bool
        The boolean value to serialize.

    Returns
    -------
    str
        "true" or "false"
    """
    return "true" if value else "false"


def deserialize_bool(value: str | list[str]) -> bool:
    """Deserialize a query param string to a boolean value.

    Accepts common boolean string representations (case-insensitive):
    - true, 1, yes, on -> True
    - false, 0, no, off -> False

    Parameters
    ----------
    value : str or list[str]
        The query param value. If a list, uses the last value.

    Returns
    -------
    bool
        The deserialized boolean value.
    """
    if isinstance(value, list):
        value = value[-1] if value else "false"

    return value.lower() in ("true", "1", "yes", "on")


def serialize_string(value: str | None) -> str:
    """Serialize a string value to a query param string.

    Parameters
    ----------
    value : str or None
        The string value to serialize.

    Returns
    -------
    str
        The string value, or empty string if None.
    """
    return value if value is not None else ""


def deserialize_string(value: str | list[str]) -> str:
    """Deserialize a query param string to a string value.

    Parameters
    ----------
    value : str or list[str]
        The query param value. If a list, uses the last value.

    Returns
    -------
    str
        The deserialized string value.
    """
    if isinstance(value, list):
        return value[-1] if value else ""
    return value


def serialize_number(value: int | float | None) -> str:
    """Serialize a number value to a query param string.

    Uses appropriate precision to avoid floating point representation issues.

    Parameters
    ----------
    value : int, float, or None
        The number value to serialize.

    Returns
    -------
    str
        The string representation of the number.
    """
    if value is None:
        return ""

    if isinstance(value, int):
        return str(value)

    # For floats, use repr for full precision, then clean up trailing zeros
    # but keep at least one decimal place for clarity
    result = repr(value)
    # Handle cases like "1.0" -> keep as "1.0", but "1.10" -> "1.1"
    if "." in result and "e" not in result.lower():
        # Remove trailing zeros, but keep at least one digit after decimal
        parts = result.split(".")
        decimal_part = parts[1].rstrip("0") or "0"
        result = f"{parts[0]}.{decimal_part}"
    return result


def deserialize_number(
    value: str | list[str], as_int: bool = False
) -> int | float | None:
    """Deserialize a query param string to a number value.

    Parameters
    ----------
    value : str or list[str]
        The query param value. If a list, uses the last value.
    as_int : bool
        If True, parse as integer. Otherwise, parse as float.

    Returns
    -------
    int, float, or None
        The deserialized number value, or None if empty/invalid.
    """
    if isinstance(value, list):
        value = value[-1] if value else ""

    if not value:
        return None

    try:
        if as_int:
            # Handle float strings by converting to float first, then int
            return int(float(value))
        return float(value)
    except ValueError:
        return None


def serialize_date(value: date | None) -> str:
    """Serialize a date value to ISO 8601 format.

    Parameters
    ----------
    value : date or None
        The date value to serialize.

    Returns
    -------
    str
        ISO 8601 date string (YYYY-MM-DD), or empty string if None.
    """
    if value is None:
        return ""
    return value.isoformat()


def deserialize_date(value: str | list[str]) -> date | None:
    """Deserialize an ISO 8601 date string to a date value.

    Parameters
    ----------
    value : str or list[str]
        The query param value. If a list, uses the last value.

    Returns
    -------
    date or None
        The deserialized date value, or None if empty/invalid.
    """
    if isinstance(value, list):
        value = value[-1] if value else ""

    if not value:
        return None

    try:
        return date.fromisoformat(value)
    except ValueError:
        return None


def serialize_date_range(value: tuple[date, date] | None) -> str:
    """Serialize a date range to comma-separated ISO 8601 format.

    Parameters
    ----------
    value : tuple[date, date] or None
        The date range (start, end) to serialize.

    Returns
    -------
    str
        Comma-separated ISO dates (YYYY-MM-DD,YYYY-MM-DD), or empty if None.
    """
    if value is None or len(value) != 2:
        return ""
    return f"{value[0].isoformat()},{value[1].isoformat()}"


def deserialize_date_range(value: str | list[str]) -> tuple[date, date] | None:
    """Deserialize comma-separated ISO 8601 dates to a date range.

    Parameters
    ----------
    value : str or list[str]
        The query param value. If a list, uses the last value.

    Returns
    -------
    tuple[date, date] or None
        The deserialized date range, or None if empty/invalid.
    """
    if isinstance(value, list):
        value = value[-1] if value else ""

    if not value:
        return None

    try:
        parts = value.split(",")
        if len(parts) != 2:
            return None
        return (date.fromisoformat(parts[0]), date.fromisoformat(parts[1]))
    except ValueError:
        return None


def serialize_date_input_value(value: Any) -> str:
    """Serialize a st.date_input value to a query param string.

    Notes
    -----
    `st.date_input` can return either a single date (scalar) or a tuple of up to
    two dates (range). For ranges, we serialize as comma-separated ISO dates.
    """
    if value is None:
        return ""

    # Range: tuple/list of dates
    if isinstance(value, (tuple, list)):
        if len(value) == 0:
            return ""
        if len(value) == 1:
            return serialize_date(value[0])
        if len(value) == 2:
            return serialize_date_range((value[0], value[1]))
        return ""

    if isinstance(value, datetime):
        return serialize_date(value.date())
    if isinstance(value, date):
        return serialize_date(value)

    return ""


def deserialize_date_input_value(value: str | list[str]) -> Any:
    """Deserialize a query param string into a st.date_input value.

    Returns either:
    - `datetime.date` for scalar values
    - `tuple[datetime.date, ...]` for comma-separated ranges (0-2 items)
    - `None` for empty/invalid values
    """
    if isinstance(value, list):
        value = value[-1] if value else ""

    if not value:
        return None

    if "," in value:
        parts = [p for p in value.split(",") if p]
        if len(parts) == 0 or len(parts) > 2:
            return None
        try:
            return tuple(date.fromisoformat(p) for p in parts)
        except ValueError:
            return None

    # Scalar date
    return deserialize_date(value)


def serialize_slider_value(value: Any) -> str:
    """Serialize a st.slider value to a query param string.

    Notes
    -----
    Sliders can return scalars (int/float/date/time/datetime) or 2-tuples for ranges.
    For ranges, we serialize as comma-separated values.
    """
    if value is None:
        return ""

    def serialize_scalar(v: Any) -> str:
        if isinstance(v, datetime):
            return serialize_datetime(v)
        if isinstance(v, date) and not isinstance(v, datetime):
            return serialize_date(v)
        if isinstance(v, time):
            return serialize_time(v)
        if isinstance(v, (int, float)):
            return serialize_number(v)
        return ""

    if isinstance(value, (tuple, list)):
        if len(value) != 2:
            return ""
        left = serialize_scalar(value[0])
        right = serialize_scalar(value[1])
        if not left or not right:
            return ""
        return f"{left},{right}"

    return serialize_scalar(value)


def deserialize_slider_value(
    value: str | list[str],
    *,
    data_type: str,
) -> Any:
    """Deserialize a query param string into a st.slider value.

    Parameters
    ----------
    value : str or list[str]
        Query param value (uses last item if a list).
    data_type : str
        One of: "int", "float", "date", "time", "datetime".

    Returns
    -------
    Any
        Scalar value for non-range sliders, or a tuple for range sliders. Returns
        None if empty/invalid.
    """
    if isinstance(value, list):
        value = value[-1] if value else ""

    if not value:
        return None

    def parse_scalar(raw: str) -> Any:
        if data_type == "int":
            parsed = deserialize_number(raw, as_int=True)
            return parsed if isinstance(parsed, int) else None
        if data_type == "float":
            return deserialize_number(raw, as_int=False)
        if data_type == "date":
            return deserialize_date(raw)
        if data_type == "time":
            return deserialize_time(raw)
        if data_type == "datetime":
            return deserialize_datetime(raw)
        return None

    if "," in value:
        parts = value.split(",")
        if len(parts) != 2:
            return None
        left = parse_scalar(parts[0])
        right = parse_scalar(parts[1])
        if left is None or right is None:
            return None
        return (left, right)

    return parse_scalar(value)


def serialize_time(value: time | None) -> str:
    """Serialize a time value to ISO 8601 format.

    Parameters
    ----------
    value : time or None
        The time value to serialize.

    Returns
    -------
    str
        ISO 8601 time string (HH:MM:SS or HH:MM), or empty string if None.
    """
    if value is None:
        return ""
    # Use isoformat which gives HH:MM:SS or HH:MM:SS.ffffff
    result = value.isoformat()
    # Strip microseconds if present and zero
    if "." in result:
        result = result.split(".")[0]
    return result


def deserialize_time(value: str | list[str]) -> time | None:
    """Deserialize an ISO 8601 time string to a time value.

    Parameters
    ----------
    value : str or list[str]
        The query param value. If a list, uses the last value.

    Returns
    -------
    time or None
        The deserialized time value, or None if empty/invalid.
    """
    if isinstance(value, list):
        value = value[-1] if value else ""

    if not value:
        return None

    try:
        return time.fromisoformat(value)
    except ValueError:
        return None


def serialize_datetime(value: datetime | None) -> str:
    """Serialize a datetime value to ISO 8601 format.

    Parameters
    ----------
    value : datetime or None
        The datetime value to serialize.

    Returns
    -------
    str
        ISO 8601 datetime string, or empty string if None.
    """
    if value is None:
        return ""
    # Use isoformat for standard representation
    result = value.isoformat()
    # Strip microseconds if zero
    if "." in result:
        result = result.split(".")[0]
    return result


def deserialize_datetime(value: str | list[str]) -> datetime | None:
    """Deserialize an ISO 8601 datetime string to a datetime value.

    Parameters
    ----------
    value : str or list[str]
        The query param value. If a list, uses the last value.

    Returns
    -------
    datetime or None
        The deserialized datetime value, or None if empty/invalid.
    """
    if isinstance(value, list):
        value = value[-1] if value else ""

    if not value:
        return None

    try:
        return datetime.fromisoformat(value)
    except ValueError:
        return None


def serialize_number_range(value: tuple[int | float, int | float] | None) -> str:
    """Serialize a number range to comma-separated format.

    Parameters
    ----------
    value : tuple[int|float, int|float] or None
        The number range (min, max) to serialize.

    Returns
    -------
    str
        Comma-separated numbers (min,max), or empty if None.
    """
    if value is None or len(value) != 2:
        return ""
    return f"{serialize_number(value[0])},{serialize_number(value[1])}"


def deserialize_number_range(
    value: str | list[str], as_int: bool = False
) -> tuple[int | float, int | float] | None:
    """Deserialize comma-separated numbers to a number range.

    Parameters
    ----------
    value : str or list[str]
        The query param value. If a list, uses the last value.
    as_int : bool
        If True, parse as integers. Otherwise, parse as floats.

    Returns
    -------
    tuple[int|float, int|float] or None
        The deserialized number range, or None if empty/invalid.
    """
    if isinstance(value, list):
        value = value[-1] if value else ""

    if not value:
        return None

    try:
        parts = value.split(",")
        if len(parts) != 2:
            return None
        if as_int:
            return (int(float(parts[0])), int(float(parts[1])))
        return (float(parts[0]), float(parts[1]))
    except ValueError:
        return None


def serialize_string_list(value: Sequence[str] | None) -> list[str]:
    """Serialize a list of strings for repeated query params.

    Parameters
    ----------
    value : Sequence[str] or None
        The list of strings to serialize.

    Returns
    -------
    list[str]
        The list of strings, or empty list if None.
    """
    if value is None:
        return []
    return list(value)


def deserialize_string_list(value: str | list[str]) -> list[str]:
    """Deserialize query params to a list of strings.

    Parameters
    ----------
    value : str or list[str]
        The query param value(s).

    Returns
    -------
    list[str]
        The list of string values.
    """
    if isinstance(value, list):
        return value
    return [value] if value else []


def serialize_option(
    value: Any,
    _options: Sequence[Any],
    key_func: Callable[[Any], str] | None = None,
) -> str:
    """Serialize a selected option value to a query param string.

    Uses str(value) by default, or a custom key function if provided.

    Parameters
    ----------
    value : Any
        The selected option value.
    _options : Sequence[Any]
        The available options. Currently unused but included for API consistency
        with deserialize_option.
    key_func : Callable[[Any], str] or None
        Optional function to extract a string key from an option.

    Returns
    -------
    str
        The serialized option key.
    """
    if value is None:
        return ""

    if key_func is not None:
        return str(key_func(value))
    return str(value)


def deserialize_option(
    value: str | list[str],
    options: Sequence[Any],
    key_func: Callable[[Any], str] | None = None,
    default: Any = None,
) -> Any:
    """Deserialize a query param string to a selected option value.

    Finds the option whose key matches the query param value.

    Parameters
    ----------
    value : str or list[str]
        The query param value. If a list, uses the last value.
    options : Sequence[Any]
        The available options to search.
    key_func : Callable[[Any], str] or None
        Optional function to extract a string key from an option.
    default : Any
        The default value if no match is found.

    Returns
    -------
    Any
        The matched option value, or default if not found.
    """
    if isinstance(value, list):
        value = value[-1] if value else ""

    if not value:
        return default

    for option in options:
        option_key = str(key_func(option)) if key_func else str(option)
        if option_key == value:
            return option

    return default


def serialize_multiselect(
    values: Sequence[Any] | None,
    _options: Sequence[Any],
    key_func: Callable[[Any], str] | None = None,
) -> list[str]:
    """Serialize selected multiselect values to query param strings.

    Parameters
    ----------
    values : Sequence[Any] or None
        The selected option values.
    _options : Sequence[Any]
        The available options. Currently unused but included for API consistency
        with deserialize_multiselect.
    key_func : Callable[[Any], str] or None
        Optional function to extract a string key from an option.

    Returns
    -------
    list[str]
        The serialized option keys.
    """
    if values is None:
        return []

    if key_func is not None:
        return [str(key_func(val)) for val in values]
    return [str(val) for val in values]


def deserialize_multiselect(
    values: str | list[str],
    options: Sequence[Any],
    key_func: Callable[[Any], str] | None = None,
) -> list[Any]:
    """Deserialize query param strings to selected multiselect values.

    Parameters
    ----------
    values : str or list[str]
        The query param value(s).
    options : Sequence[Any]
        The available options to search.
    key_func : Callable[[Any], str] or None
        Optional function to extract a string key from an option.

    Returns
    -------
    list[Any]
        The matched option values.
    """
    if isinstance(values, str):
        values = [values] if values else []

    result = []
    for val in values:
        for option in options:
            option_key = str(key_func(option)) if key_func else str(option)
            if option_key == val:
                result.append(option)
                break
    return result


def serialize_color(value: str | None) -> str:
    """Serialize a hex color value to a query param string.

    Removes the leading '#' for cleaner URLs.

    Parameters
    ----------
    value : str or None
        The hex color value (e.g., "#ff0000").

    Returns
    -------
    str
        The hex color without '#' (e.g., "ff0000").
    """
    if value is None:
        return ""
    return value.lstrip("#").lower()


def deserialize_color(value: str | list[str]) -> str | None:
    """Deserialize a query param string to a hex color value.

    Adds the leading '#' if not present.

    Parameters
    ----------
    value : str or list[str]
        The query param value. If a list, uses the last value.

    Returns
    -------
    str or None
        The hex color with '#' (e.g., "#ff0000"), or None if empty.
    """
    if isinstance(value, list):
        value = value[-1] if value else ""

    if not value:
        return None

    # Add '#' prefix if not present
    if not value.startswith("#"):
        value = f"#{value}"

    return value.lower()
